feat(chat): add UI option to mute group peers

tox-user opened the pull request:
Implements: #4596, #4626, #1007
Adds a context menu to every peer label in group chat with an option to mute chosen peers, which adds them to blacklist.

![qtox mute peer 2](https://user-images.githubusercontent.com/32248732/39990990-9e5a8fd8-575d-11e8-8f5a-a9dff8f72432.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/5124)
<!-- Reviewable:end -->

sudden6 commented:
Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---

*[src/widget/form/groupchatform.cpp, line 251 at r1](https://reviewable.io/reviews/qtox/qtox/5124#-LBqryqNEPMCKX7Ao2_I:-LBqryqNEPMCKX7Ao2_J:b-dz89x3) ([raw file](https://github.com/qtox/qtox/blob/e45d7a6f6ec4303e7f720ddd0997d5c83649b653/src/widget/form/groupchatform.cpp#L251)):*
> ```cpp
>         const ToxPk peerPk = core->getGroupPeerPk(group->getId(), peerNumber);
> 
>         connect(label, &QLabel::customContextMenuRequested, [label, peerPk, this, editedName](const QPoint& localPos) {
> ```

I think instead of writing such a big lambda, it would be better to create an extra class method. Lambdas are better suited for very short things IMO

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
I can do that, but I don't know how to connect a method to a signal in a way that would allow me to use the signal's argument `const QPoint&` and variables: `label`, `peerPk`, `editedName`. A lambda function allows me to use all of them. How do I achieve that without the lambda?

sudden6 commented:
You can just make them parameters or you can get `label` via `QObject::sender()`, get `editedName` via `label->text()`, or if you wait for my ToxPk PR to be merged, you can easily get all that by only passing `peerPk`

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124#-:-LBvbUCs3GAI1y8NfBE3:b-hw22bt)*
<!-- Sent from Reviewable.io -->

tox-user committed:
feat(chat): add UI option to mute group peers

Implements #4596, #4626, #1007

Adds a context menu to every peer label in group chat with an
option to mute chosen peers, which adds them to the blacklist.
Mutes audio of blacklisted group peers.

tox-user commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/widget/form/groupchatform.cpp, line 251 at r1](https://reviewable.io/reviews/qtox/qtox/5124#-LBqryqNEPMCKX7Ao2_I:-LCT5kxY0RL7Kv1RjII4:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/fab96c4774d46259c58361754d22c1eefd914584/src/widget/form/groupchatform.cpp#L251)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

I think instead of writing such a big lambda, it would be better to create an extra class method. Lambdas are better suited for very short things IMO
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
I'm not entirely sure, but could there be a resource leak with `contextMenu`? It seems to get created every time I do a right click, but never destroyed? Or does it destroy itself after it's closed?

tox-user commented:
How can I test that to find out?

sudden6 commented:
check if the constructor of that menu object is ever called (not sure if that's easy), use valgrind or watch memory usage while you trigger the menu constantly

tox-user commented:
Creating a lot of those menus does increase memory usage by a small amount. What should I do?

sudden6 commented:
Create the context menu only once and store a pointer to it somewhere we can find it again or find a way to call the destructor once it's not needed anymore.

tox-user commented:
Fixed.

sudden6 commented:
Reviewed 3 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---

*[src/widget/form/groupchatform.cpp, line 510 at r2](https://reviewable.io/reviews/qtox/qtox/5124#-LCof0XPA7sozsgXd8DN:-LCof0XPA7sozsgXd8DO:bs0b8v8) ([raw file](https://github.com/qtox/qtox/blob/adc7b55506f220b142047a9d1c06b267e0a561ab/src/widget/form/groupchatform.cpp#L510)):*
> ```cpp
> 
>     // delete menu after it stops being used
>     contextMenu->deleteLater();
> ```

Are you sure this doesn't delete the menu too early? What do you think about connecting this to the `aboutToHide()` signal? I found this solution also here: https://stackoverflow.com/questions/39259862/when-and-how-to-properly-destroy-qmenu-context-menu#comment81777174_39260237

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
I thought it would close too early too, but I tested it and it doesn't. I don't know why, but it works :D. I couldn't get `aboutToHide()` to work. The signal never fired.

tox-user commented:
Actually it does work when I connect it earlier in the code. I'm gonna do that then.

tox-user commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/widget/form/groupchatform.cpp, line 510 at r2](https://reviewable.io/reviews/qtox/qtox/5124#-LCof0XPA7sozsgXd8DN:-LCownag8wRnhfl0XYvI:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/adc7b55506f220b142047a9d1c06b267e0a561ab/src/widget/form/groupchatform.cpp#L510)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Are you sure this doesn't delete the menu too early? What do you think about connecting this to the `aboutToHide()` signal? I found this solution also here: https://stackoverflow.com/questions/39259862/when-and-how-to-properly-destroy-qmenu-context-menu#comment81777174_39260237
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
@tox-user forgot to push?

tox-user commented:
Yes, sorry.

sudden6 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124#-:-LCp91Fu-QwxT1XuJKFI:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDR1-7H0ecJuib30sBG:bh8qlb2) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
> ```cpp
> 
> namespace
> {
> ```

Maybe it would be worth to have inner `consts` namespace to make these variables look less-globalish? I mean:
```
namespace 
{
namespace consts 
{
...
``
```


---

*[src/widget/form/groupchatform.cpp, line 49 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1LWqBXoAJxWGe7ws:-LDR1LWqBXoAJxWGe7wt:bbb5cd5) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L49)):*
> ```cpp
> namespace
> {
> const auto LABEL_PEER_TYPE_OUR = QVariant("our");
> ```

maybe QStringLiteral for string literals?

---

*[src/widget/form/groupchatform.cpp, line 450 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2HyM3cFzxjPizedn:-LDR2HyM3cFzxjPizedo:b44trbt) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L450)):*
> ```cpp
> void GroupChatForm::onLabelContextMenuRequested(const QPoint& localPos)
> {
>     QLabel* label = static_cast<QLabel*>(QObject::sender());
> ```

QLabel* const label

---

*[src/widget/form/groupchatform.cpp, line 460 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2WWDCLNg0Mndf_Yp:-LDR2WWDCLNg0Mndf_Yq:b8xwzby) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L460)):*
> ```cpp
>     const QString unmuteString = tr("unmute");
>     Settings& s = Settings::getInstance();
>     QStringList blackList = s.getBlackList();
> ```

const QStringList

---

*[src/widget/form/groupchatform.cpp, line 461 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2_VZ-70U5rtuPSrt:-LDR2_VZ-70U5rtuPSru:b-7kvz78) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L461)):*
> ```cpp
>     Settings& s = Settings::getInstance();
>     QStringList blackList = s.getBlackList();
>     QMenu* contextMenu = new QMenu(this);
> ```

const contextMenu

---

*[src/widget/form/groupchatform.cpp, line 469 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3PK20dYzB9hsHrMS:-LDR3PK20dYzB9hsHrMT:b-6l9jwb) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L469)):*
> ```cpp
>     connect(contextMenu, &QMenu::aboutToHide, contextMenu, &QObject::deleteLater);
> 
>     QMap<ToxPk, QLabel*>::const_iterator i = peerLabels.constBegin();
> ```

Maybe use std::find_if, it would be much clearer from code reader part. Also, probably no need to have `peerPk` variable as resulted iterator could be used directly then?

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDR49wm8xgJGZ4XudF_:b-u1b6s5) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
> ```cpp
> 
>     const bool isPeerBlocked = blackList.contains(peerPk.toString());
>     QString menuTitle = label->text();
> ```


Also, using lambda and cost could bring in cleariness, as this non-const variable will not actually be changed later:

const QString menuTitle = [label] () {

}();

As for https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init

---

*[src/widget/form/groupchatform.cpp, line 484 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3i3iAxNL9sFkvaCE:-LDR3i3iAxNL9sFkvaCF:b-a86v79) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L484)):*
> ```cpp
>     const bool isPeerBlocked = blackList.contains(peerPk.toString());
>     QString menuTitle = label->text();
>     if (menuTitle.endsWith(", ")) {
> ```

QLatin1String(", ")

---

*[src/widget/form/groupchatform.cpp, line 492 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR4PKnAUttngCUvnbn:-LDR4PKnAUttngCUvnbo:bmr02hb) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L492)):*
> ```cpp
> 
>     if (isPeerBlocked) {
>         toggleMuteAction = contextMenu->addAction(unmuteString);
> ```

toogleMuteAction variable is declared wayy to much too early, imho. Could be also const & lambda init.

---

*[src/widget/form/groupchatform.cpp, line 501 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR4f7LDpbGg7g77fCX:-LDR4f7LDpbGg7g77fCY:b-liu8qb) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L501)):*
> ```cpp
>     if (selectedItem == toggleMuteAction) {
>         if (isPeerBlocked) {
>             int index = blackList.indexOf(peerPk.toString());
> ```

const int

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDbHNGM8gbDu7Vs8Lj_:bb76mu4) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

Maybe it would be worth to have inner `consts` namespace to make these variables look less-globalish? I mean:
```
namespace 
{
namespace consts 
{
...
```
</blockquote></details>

Why?

---

*[src/widget/form/groupchatform.cpp, line 49 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1LWqBXoAJxWGe7ws:-LDbHRvnAT-J82W-qCLL:bsa4b3y) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L49)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

maybe QStringLiteral for string literals?
</blockquote></details>

Even when a string is just a simple word with no special characters? Why?

---

*[src/widget/form/groupchatform.cpp, line 450 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2HyM3cFzxjPizedn:-LDbLRNE40JajcElrQUJ:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L450)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

QLabel* const label
</blockquote></details>

Done.

---

*[src/widget/form/groupchatform.cpp, line 460 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2WWDCLNg0Mndf_Yp:-LDbKfKA0AxrQkXt38w7:b-s42770) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L460)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const QStringList
</blockquote></details>

Results in a compilation error:
```
groupchatform.cpp:503:41: error: passing ‘const QStringList’ as ‘this’ argument discards qualifiers [-fpermissive]
                 blackList.removeAt(index);
```

---

*[src/widget/form/groupchatform.cpp, line 461 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2_VZ-70U5rtuPSrt:-LDbM3tz6MdQdTCLT75b:b-cjfs25) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L461)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const contextMenu
</blockquote></details>

Results in a compilation error:
```
groupchatform.cpp:487:64: error: no matching function for call to ‘QMenu::addAction(QString&) const’
     QAction* menuTitleAction = contextMenu->addAction(menuTitle);
```

---

*[src/widget/form/groupchatform.cpp, line 469 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3PK20dYzB9hsHrMS:-LDbQ9ujBl3yvZlF3GhE:b-76jbj6) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L469)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

Maybe use std::find_if, it would be much clearer from code reader part. Also, probably no need to have `peerPk` variable as resulted iterator could be used directly then?
</blockquote></details>

Could you give an example?

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDbJQJr7V0FFzsSeywz:b-99zvic) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>


Also, using lambda and cost could bring in cleariness, as this non-const variable will not actually be changed later:

const QString menuTitle = [label] () {

}();

As for https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init
</blockquote></details>

It will make code harder to read.

---

*[src/widget/form/groupchatform.cpp, line 484 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3i3iAxNL9sFkvaCE:-LDbI-SECHTMhY-JEagR:bb76mu4) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L484)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

QLatin1String(", ")
</blockquote></details>

Why?

---

*[src/widget/form/groupchatform.cpp, line 492 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR4PKnAUttngCUvnbn:-LDbNrn6C-OVb0ckxCXi:b-khpof4) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L492)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

toogleMuteAction variable is declared wayy to much too early, imho. Could be also const & lambda init.
</blockquote></details>

I moved it down.

---

*[src/widget/form/groupchatform.cpp, line 501 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR4f7LDpbGg7g77fCX:-LDbNuBE-M7oso7HJEyz:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L501)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

const int
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDbU7-s5jc-UYiB9qSj:b-2d5gj2) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Why?
</blockquote></details>

better for autocompletion :) , also unscoped name may theoretically clash with some other global, if any... But that's mostly taste issue I guess.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[src/widget/form/groupchatform.cpp, line 49 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1LWqBXoAJxWGe7ws:-LDbUZDB6IMm_h8ZM9Gg:b-yzu31i) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L49)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Even when a string is just a simple word with no special characters? Why?
</blockquote></details>

Because on every exection, QString(char*) constructor will be called, and every time it will have to calculate string lenght, allocate buffer and copy "our" into that new buffer. QStringLitera() macro will return "baked" constant QString.

---

*[src/widget/form/groupchatform.cpp, line 460 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2WWDCLNg0Mndf_Yp:-LDbUq0vBN9UlfbhckzB:bjhczvx) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L460)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Results in a compilation error:
```
groupchatform.cpp:503:41: error: passing ‘const QStringList’ as ‘this’ argument discards qualifiers [-fpermissive]
                 blackList.removeAt(index);
```
</blockquote></details>

Ok, missed that mutable call.

---

*[src/widget/form/groupchatform.cpp, line 461 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2_VZ-70U5rtuPSrt:-LDbUvJnEKo9CUkQWMHI:b3dvjwz) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L461)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Results in a compilation error:
```
groupchatform.cpp:487:64: error: no matching function for call to ‘QMenu::addAction(QString&) const’
     QAction* menuTitleAction = contextMenu->addAction(menuTitle);
```
</blockquote></details>

I had in mind `QMenu* const contextMenu`, not `const QMenu* contextMenu`. Object is mutable, yes, but pointer itself is not changed.

---

*[src/widget/form/groupchatform.cpp, line 469 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3PK20dYzB9hsHrMS:-LDbV9V88gtJBz8JyMqz:b-la8pxv) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L469)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Could you give an example?
</blockquote></details>

```
const peerPkIt = std::find_if(peerLabels.constBegin(), peerLabels.constEnd(), [label](QLabel* item) {
    retrun (item == label);
});

if (peerPkIt == peerLabels.constEnd() || peerPkIt.value() == selfPk) {
    return;
}

// ...

const bool isPeerBlocked = blackList.contains(peerPkIt.value().toString());
```

Though actually, there is member `QMap::key()` that returns key by value, so that  would be even simpler, no need for loop neither for algorithm :) .

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDbXu25CI22Giee50uw:b-uztyf) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

It will make code harder to read.
</blockquote></details>

It will make easier to see that this variable is no longer modified and you can unload  it from your brain cache :) .

---

*[src/widget/form/groupchatform.cpp, line 484 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3i3iAxNL9sFkvaCE:-LDbY0Lv53VnXJE3qHCE:b-b7vlna) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L484)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Why?
</blockquote></details>

See https://github.com/qTox/qTox/blob/master/doc/coding_standards.md#strings and links below for best Qt "stringing" practices.

In this case, "prefer QLatin1String overload if it exists". 

Looping and comparing against 8-bit string is faster than to utf-16 one (QSTring default), and no QString temporary is created in this case at all.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDbh9_gB5RJUGDHlAvz:bpdilmb) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

better for autocompletion :) , also unscoped name may theoretically clash with some other global, if any... But that's mostly taste issue I guess.
</blockquote></details>

We don't do that anywhere, so I will keep it as is.

---

*[src/widget/form/groupchatform.cpp, line 49 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1LWqBXoAJxWGe7ws:-LDbh0zg6Bkiwde-lngA:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L49)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

Because on every exection, QString(char*) constructor will be called, and every time it will have to calculate string lenght, allocate buffer and copy "our" into that new buffer. QStringLitera() macro will return "baked" constant QString.
</blockquote></details>

Done.

---

*[src/widget/form/groupchatform.cpp, line 461 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR2_VZ-70U5rtuPSrt:-LDbgzQcByKdWYK63A6J:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L461)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

I had in mind `QMenu* const contextMenu`, not `const QMenu* contextMenu`. Object is mutable, yes, but pointer itself is not changed.
</blockquote></details>

Done.

---

*[src/widget/form/groupchatform.cpp, line 469 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3PK20dYzB9hsHrMS:-LDbgh9s-d1sJsOboBel:b-tpkkjl) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L469)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

```
const peerPkIt = std::find_if(peerLabels.constBegin(), peerLabels.constEnd(), [label](QLabel* item) {
    retrun (item == label);
});

if (peerPkIt == peerLabels.constEnd() || peerPkIt.value() == selfPk) {
    return;
}

// ...

const bool isPeerBlocked = blackList.contains(peerPkIt.value().toString());
```

Though actually, there is member `QMap::key()` that returns key by value, so that  would be even simpler, no need for loop neither for algorithm :) .

</blockquote></details>

Done. Much better now :).

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDbiMRR-UjVq_eKM4Wj:bqmw0mf) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

It will make easier to see that this variable is no longer modified and you can unload  it from your brain cache :) .
</blockquote></details>

Then I have to load another function to my cache :). Doesn't seem worth it. If we did that often, it would be terrible.

---

*[src/widget/form/groupchatform.cpp, line 484 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR3i3iAxNL9sFkvaCE:-LDbgdrw7uTRSQiL8wuN:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L484)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

See https://github.com/qTox/qTox/blob/master/doc/coding_standards.md#strings and links below for best Qt "stringing" practices.

In this case, "prefer QLatin1String overload if it exists". 

Looping and comparing against 8-bit string is faster than to utf-16 one (QSTring default), and no QString temporary is created in this case at all.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDgiA9kBPbnQ460ttva:b-5d3ca7) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

We don't do that anywhere, so I will keep it as is.
</blockquote></details>

https://github.com/qTox/qTox/blob/5dc4e6de81690d51e9ce92f7af264099b891ee3c/src/net/toxmedata.cpp#L29

I wound't say that's critical, but maybe it's a good habit to organize variables better.

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDgmpLO3bEqWLK40rAK:b-rsw1bw) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

Then I have to load another function to my cache :). Doesn't seem worth it. If we did that often, it would be terrible.
</blockquote></details>

It's more like a scope than a function, as it does not have a name and is written in-place.

There is kinda a nano-movement to "const as much as possible", and perform complex initialization via lambdas:

https://www.youtube.com/watch?v=d4nmNYTM1j8&feature=youtu.be

https://herbsutter.com/2013/04/05/complex-initialization-for-a-const-variable/

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es28-use-lambdas-for-complex-initialization-especially-of-const-variables

Of course, this particular case is not *hat* complex. But in other cases one can also avoid keeping more compex variables for long time, for example:

```
const int port = []() {
   BigSettingsClass settings;
   retrun settings.getPort();
  // BigSettingsClass no needed for anything else in this function will be destroyed here. 
}();
```
But again, we can skip this particular case...

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDmAh4E3rYxydj24rBZ:b-hnhnav) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

https://github.com/qTox/qTox/blob/5dc4e6de81690d51e9ce92f7af264099b891ee3c/src/net/toxmedata.cpp#L29

I wound't say that's critical, but maybe it's a good habit to organize variables better.
</blockquote></details>

It's better, because your IDE will show it better? Everyone uses different tools.

---

*[src/widget/form/groupchatform.cpp, line 483 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR49wl1QSOcLs1P01S:-LDmAaeI6GfSm0enZ4ca:bxpmrxm) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L483)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

It's more like a scope than a function, as it does not have a name and is written in-place.

There is kinda a nano-movement to "const as much as possible", and perform complex initialization via lambdas:

https://www.youtube.com/watch?v=d4nmNYTM1j8&feature=youtu.be

https://herbsutter.com/2013/04/05/complex-initialization-for-a-const-variable/

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es28-use-lambdas-for-complex-initialization-especially-of-const-variables

Of course, this particular case is not *hat* complex. But in other cases one can also avoid keeping more compex variables for long time, for example:

```
const int port = []() {
   BigSettingsClass settings;
   retrun settings.getPort();
  // BigSettingsClass no needed for anything else in this function will be destroyed here. 
}();
```
But again, we can skip this particular case...
</blockquote></details>

I think it's a really bad idea. If we did that all the time, our code would be much harder to read. It would be as bad as overusing auto. Maybe there is a gain in performance, but it doesn't seem worth it to me. Perhaps there exists a rare case, where creating a lambda for variables is required, but doing it every time seems bad.

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

Talkless commented:
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/widget/form/groupchatform.cpp, line 48 at r3](https://reviewable.io/reviews/qtox/qtox/5124#-LDR1-7H0ecJuib30sBF:-LDr14nq6a6Iy0CF4DGr:b-dyariu) ([raw file](https://github.com/qtox/qtox/blob/6b12ab4d22e258bc4016bf63c074a3eefc096f09/src/widget/form/groupchatform.cpp#L48)):*
<details><summary><i>Previously, tox-user wrote…</i></summary><blockquote>

It's better, because your IDE will show it better? Everyone uses different tools.
</blockquote></details>

Not only better for IDE, but your own eyes. There are some member static variables in qTox classes - these are without extra scope name, obviously. Meanwhile for "global" constants namespaces show where they belong. `consts` could be used for local conststants, other namespaces for other purposes, and that fixes ambiguities. 

Talking about possible ambiguities, [names should not be upper cased](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es9-avoid-all_caps-names), as these are kinda "reserved" for C processor macros, that does not know about scopes at all. Though C-isms is deep into our culture I guess :) .

---


*Comments from [Reviewable](https://reviewable.io/reviews/qtox/qtox/5124)*
<!-- Sent from Reviewable.io -->

tox-user commented:
I am following qTox's code guidelines and use mostly the same style that is used by other qTox developers. If you disagree with the guidelines or have ideas for improvements, you should create an issue about it. It's not productive to discuss it here, because it's something that more people would have to agree on and it might require a longer discussion. If everyone approves those ideas, we will then start using them and try to change the old code to do the same. Until then I will keep doing things the same way.

Diadlo commented:
@tox-user Just IMO. qTox have not the best codebase. Style and especially architecture.
@Talkless arguments mostly based not on their own opinion, but on CppCoreGuidelines, with was based on codeing expirience

tox-user commented:
It is an opinion. CppCoreGuidelines provide some arguments and many of them make sense, but I disagree with some of them.

In my opinion performance is important, but only when it has a big enough impact, like when we open and read a lot of style sheets. I try not to worry about the smallest differences in performance though.

sudden6 commented:
Could we please move the discussion of Coding Style and Guidelines to an extra issue? This has not much to do with this PR and AFAICT the PR itself is following the current guidelines.

Talkless commented:
> Could we please move the discussion of Coding Style and Guidelines to an extra issue? This has not much to do with this PR and AFAICT the PR itself is following the current guidelines.

True, it is not productive to go too deep into styles during reviews, if that style is not documented in Coding Style.
